codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java

```diff
@@ -0,0 +1,252 @@
+
+    private State state(Http2Stream stream) {
+        return checkNotNull(stream, "stream").getProperty(stateKey);
+    }
+
+    private class EnqueueStrategy {
+        void enqueue(State state) {
+            queue.addLast(state);
+            state.enqueued = true;
+        }
+    }
+
+    private final EnqueueStrategy defaultEnqueueStrategy = new EnqueueStrategy();
+
+    /**
+     * The first time a state is enqueued, it will be added to the head of the queue. After
+     * that it will be added to the tail.
+     */
+    private final EnqueueStrategy firstEnqueueStrategy = new EnqueueStrategy() {
+        @Override
+        void enqueue(State state) {
+            queue.addFirst(state);
+            state.enqueued = true;
+
+            // Switch to the default strategy so that it will be added to the tail from
+            // now on.
+            state.enqueueStrategy = defaultEnqueueStrategy;
+        }
+    };
+
+    /**
+     * The remote flow control state for a single stream.
+     */
+    private final class State {
+        final Http2Stream stream;
+        boolean hasFrame;
+        int streamableBytes;
+        int allocated;
+        boolean enqueued;
+        EnqueueStrategy enqueueStrategy = firstEnqueueStrategy;
+
+        State(Http2Stream stream) {
+            this.stream = stream;
+        }
+
+        void allocate(int bytes) {
+            allocated += bytes;
+        }
+
+        int unallocated() {
+            return streamableBytes - allocated;
+        }
+
+        /**
+         * Reset the number of bytes that have been allocated to this stream by the priority
+         * algorithm.
+         */
+        void resetAllocated() {
+            allocated = 0;
+        }
+
+        void updateStreamableBytes(int newStreamableBytes, boolean hasFrame) {
+            this.hasFrame = hasFrame;
+
+            int delta = newStreamableBytes - streamableBytes;
+            if (delta != 0) {
+                streamableBytes = newStreamableBytes;
+                totalStreamableBytes += delta;
+
+                if (!enqueued && streamableBytes > 0) {
+                    // It's not in the queue but it has streamable bytes - enqueue it.
+                    enqueue();
+                }
+            }
+        }
+
+        void enqueue() {
+            enqueueStrategy.enqueue(this);
+        }
+
+        void close() {
+            // Clear the streamable bytes.
+            updateStreamableBytes(0, false);
+
+            // Remove this state from the queue.
+            if (enqueued) {
+                queue.remove(this);
+            }
+        }
+
+        boolean hasFrame() {
+            return hasFrame;
+        }
+    }
+
+    /**
+     * A connection stream visitor that delegates to the user provided visitor.
+     */
+    private class WriteVisitor implements Http2StreamVisitor {
+        Writer writer;
+
+        void writeAllocatedBytes(Writer writer) throws Http2Exception {
+            this.writer = writer;
```

Comment:
Should we add re-entry protection like in `PriorityStreamByteDistributor`?

Also it seems like this `WriteVisitor` and `PriorityStreamByteDistributor.WriteVisitor` are doing the same thing. Is it worth sharing code with a package private class, considering the `State` type is different?